File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        from collections import deque 

      
        def bfs(x,y ):

            queue = deque()
            queue.append((x,y))
            dx = [1,-1,0,0] 
            dy = [0,0,1,-1]
            grid[x][y] ='#'
            cnt = 0 
            while queue:
                x, y  = queue.popleft()
                cnt +=1 
                for i in range(4):
                    nx = x +dx[i]
                    ny = y +dy[i]
                    if nx <0 or nx >=R or ny<0 or ny>=C or grid[nx][ny] !=1:
                        continue 
                    if grid[nx][ny] ==1:
                        queue.append((nx,ny))
                        grid[nx][ny] ='#'
            return cnt 

        R = len(grid) ; C = len(grid[0])
        ans =0 
        for r  in range(R):
            for c in range(C):
                if grid[r][c] ==1:
                    ans = max(ans, bfs(r,c))

        return ans 

File: 695-max-area-of-island/695-max-area-of-island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& mat) {
        int ans = 0 ; 
        int R = mat.size() , C = mat[0].size() ; 

        for(int r=0 ; r<R ; r++){
            for(int c=0 ; c<C ; c++){ 
                if( mat[r][c] ==1 ){ 
                    ans = max(ans, dfs(r,c, mat)) ;;  
        }}}
        return ans ; 
    }

    int dfs(int x,int y, vector<vector<int>>& mat){ 
        int R = mat.size() , C = mat[0].size() ; 
        vector<vector<int>> dir {{1,0},{-1,0},{0,1},{0,-1}} ; 
        mat[x][y] = 0 ; 
        int cnt = 1; 
        for(auto d:dir){ 
            int nx = x+d[0] , ny = y+d[1] ; 
            if(0<=nx && nx< R && 0<=ny && ny<C)
                if(mat[nx][ny] ==1)
                    cnt +=  dfs(nx,ny ,mat ) ; } 
        return cnt ;
    }

};